protostar/configuration_file/configuration_toml_reader.py

```diff
@@ -0,0 +1,115 @@
+from pathlib import Path
+from typing import Any, Dict, List, Optional, cast
+
+import flatdict
+import tomli
+
+from protostar.protostar_exception import ProtostarException
+
+
+class ConfigurationTOMLReader:
+    FlattenSectionName = str
+    """e.g. `profile.ci.protostar.shared_command_configs`"""
+
+    def __init__(
+        self,
+        path: Path,
+    ):
+        self.path = path
+        self._cache: Optional[
+            Dict[ConfigurationTOMLReader.FlattenSectionName, Any]
+        ] = None
+
+    def get_filename(self) -> str:
+        return self.path.name
+
+    def get_section(
+        self,
+        section_name: str,
+        profile_name: Optional[str] = None,
+        section_namespace: Optional[str] = None,
+    ) -> Optional[Dict[str, Any]]:
+
+        section_name = (
+            f"{section_namespace}.{section_name}" if section_namespace else section_name
+        )
+
+        protostar_toml_dict = self._read_if_cache_miss()
+
+        if profile_name:
+            section_name = f"profile.{profile_name}.{section_name}"
+
+        if section_name not in protostar_toml_dict:
+            return None
+
+        return protostar_toml_dict[section_name]
+
+    def get_attribute(
+        self,
+        section_name: str,
+        attribute_name: str,
+        profile_name: Optional[str] = None,
+        section_namespace: Optional[str] = None,
+    ) -> Optional[Any]:
+        section = self.get_section(
+            section_name, profile_name, section_namespace=section_namespace
+        )
+        if not section:
+            return None
+
+        alternative_attribute_name = self._find_alternative_key(attribute_name, section)
+
+        if alternative_attribute_name and alternative_attribute_name in section:
+            return section[alternative_attribute_name]
+        return None
+
+    def get_profile_names(self) -> List[str]:
+        protostar_toml_dict = self._read_if_cache_miss()
+        section_names = list(protostar_toml_dict.keys())
+        profile_section_names = [
+            section_name
+            for section_name in section_names
+            if section_name.startswith("profile")
+        ]
+        profile_names = [
+            profile_section_name.split(".")[1]
+            for profile_section_name in profile_section_names
+        ]
+        return profile_names
+
+    @staticmethod
+    def _find_alternative_key(base_key: str, raw_dict: Dict[str, Any]) -> Optional[str]:
+        if base_key in raw_dict:
+            return base_key
+
+        underscored_variant = base_key.replace("-", "_")
+        if underscored_variant in raw_dict:
+            return underscored_variant
+
+        dashed_variant = base_key.replace("_", "-")
+        if dashed_variant in raw_dict:
+            return dashed_variant
+
+        return None
+
+    def _read_if_cache_miss(self) -> Dict[str, Any]:
```

Comment:
> why is the file lazy loaded?
To make the class composable by not putting heavy logic into the constructor.

> and in general this looks more like lazy loading than a cache (which can be evicted and would rather be per-key)

I'd say it's both. The cache logic could be moved to the `ConfigurationTOMLReaderCacheProxy` but it would work per getter which would be pointless.

What's the benefit of per-key cache in this case?

Reply:
There's no use of per-key cache here, hence I find the name `cache` misleading.

Lazy loading has a nasty properties that a problem in lazy loading is not catched in a place which called it. Would using a factory classmethod be more feasible here?

Reply:
> There's no use of per-key cache here, hence I find the name cache misleading.

Wiki definition doesn't mention that a cache must be per key:
> ...software component that stores data so that future requests for that data can be served faster; the data stored in a cache might be the result of an earlier computation or a copy of data stored elsewhere

---

> Lazy loading has a nasty properties that a problem in lazy loading is not catched in a place which called it.

I don't get it TBH.

Reply:
> I don't get it TBH.

If you have an object that fails during its construction, and you put it lazily loaded as inside some class, let's say behind a computed property; any errors will be raised when reading this property instead of when the lazy wrapper over the object is constructed.

This makes debugging harder, IMHO.

And also, but this is negligible here, this breaks the contract that `get_` operations are _fast_ -- first call will not be as it will perform I/O.

Reply:
What I pursue in many comments here is to move IO in its entirety out of this code. That way this lazy loading is easily avoidable and the simplicity is maintained.

Reply:
I'm moving IO out of this code, but I don't see how lazy loading could be avoided without affecting the resilience. 

Where should the file be opened? In the composition root? By each class that uses `ConfigurationFile`?

Reply:
I suggest opening, reading and parsing it during application startup, which should map to composition root as far as I understand.

Reply:
I removed the lazy loading. The configuration file will be used by the `ArgumentParser` anyway.